Attachments & more

teunisv opened the pull request:
Tests might fail due gmail having plain text as body part 0, and html as body part 1

For the rest open for input

Commit:
First Change

Commit:
Work Around for issue #1

Commit:
Prevent opening 100 of folders if there no messages

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
 
+    public ArrayList<String> listAttachments() {
+        ArrayList<String> values = null;
+        if(messageAttachments.size() > 0) {
```
Can you format your changes so that there is a space between `if` and `for` and their opening parentheses (`(`)?

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
+        ArrayList<String> values = null;
+        if(messageAttachments.size() > 0) {
+            values = new ArrayList<>();
+            for (ImapAttachment ia : messageAttachments) values.add(ia.getFileName());
```
Please add some braces (`{` and `}`) and spread this line across 3 lines.

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
+        if(number < messageAttachments.size()) {
+            return impSaveAttachment(messageAttachments.get(number));
+        }
+        else throw new SlimFixtureException("There are only "+ messageAttachments.size() + " attachments");
```
Spread across 3 lines.

fhoeben replied:
```suggestion
        else throw new SlimFixtureException(false, "There are only "+ messageAttachments.size() + " attachments");
```

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
+        return values;
+    }
+
+    public String saveAttachmentByNumber(int number)
```
Wouldn't `saveAttachmentNumber(` make a better command in the wiki e.g. `|show|save attachment number|1|`?

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
+        for(ImapAttachment ia: messageAttachments) {
+            if(StringUtils.equals(ia.getFileName(), filename)) return impSaveAttachment(ia);
+        }
+        throw new SlimFixtureException("There is no file with" + filename + " attachments");
```
```suggestion
        throw new SlimFixtureException(false, "There is no attachment called " + filename);
```

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
+    public String saveAttachmentByFilename(String filename)
+    {
+        for(ImapAttachment ia: messageAttachments) {
+            if(StringUtils.equals(ia.getFileName(), filename)) return impSaveAttachment(ia);
```
Spread across 3 lines

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
         });
     }
 
+    public ArrayList<String> listAttachments() {
```
Wouldn't `attachmentNames()` work better in the wiki?

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
+        else throw new SlimFixtureException("There are only "+ messageAttachments.size() + " attachments");
+    }
+
+    public String saveAttachmentByFilename(String filename)
```
`saveAttachmentNamed()`?

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
+        try {
+            return createFile(attachmentBase, ia.getFileName(), ia.getBytes());
+        } catch (IOException ex) {
+            throw new RuntimeException("Unable to readstream", ex);
```
```suggestion
            throw new SlimFixtureException("Unable to extract: " + ia.getFileName(), ex);
```

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
+                } else if (content instanceof String) {
+                    if (part.isMimeType("text/html")) messageHtml = (String)content;
+                    else messagePlain = (String)content;
+                } else throw new RuntimeException("Unknown content type " + content.getClass());
```
3 lines and throw `SlimFixtureException` with first param `false` instead

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
+                } else throw new RuntimeException("Unknown content type " + content.getClass());
+            }
+        } catch (IOException ex) {
+            throw new RuntimeException("Unable to get body of message", ex);
```
SlimFixtureException

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
+        } catch (IOException ex) {
+            throw new RuntimeException("Unable to get body of message", ex);
+        } catch (MessagingException ex) {
+            throw new RuntimeException("Unable to get body of message", ex);
```
SlimFixtureException

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
+                catch (MessagingException ex) {
+                    throw new RuntimeException("Unable to get received date", ex);
+                }
+            }).collect(Collectors.toList()).toArray();
```
Why convert this list to array?

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
-                lastMessage = messages[messages.length - 1];
-            }
-            return result;
+            Folder f = openFolder();
```
Why only close `f` if there are no messages? Shouldn't that always be done? `Folder` is `AutoClosable` so could be placed inside a `try`:
`try (Folder f = openFolder()) {`

teunisv replied:
Ill have to check if this doesn't create issues with handleParts, impSaveAttachment & moveMessage
--Edit--
This indeed creates new issues, since above assumes the folder is open.
And folder only really needs to closed if there no messages found. (the expected message still needs to arrive)

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
     }
 
     public boolean retrieveLastMessage() {
         lastMessage = null;
```
This line can be removed correct?

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
         lastMessage = null;
+        Message[] messages = retrieveMessages();
+        boolean result = (messages != null && messages.length > 0);
+        setLastMessage(result ? messages[messages.length - 1]: null  );
```
```suggestion
        setLastMessage(result? messages[messages.length - 1]: null);
```

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -105,25 +110,98 @@ public String ccRecipient() {
+            if(messages.length == 0 ) { f.close(); return null; }
+            return (Message[])Arrays.stream(messages).filter(x -> {
+                try {
+                    return x.getReceivedDate().after(receivedAfter);
```
Can receivedAfter be `null` and is this filter required then?

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -248,7 +325,46 @@ public String getFolder() {
+        String name = base + folder.getName();
+        if(!name.isEmpty()) folders.add(name);
+        for(Folder f : folder.list()) {
+            String b = name.isEmpty() ? "" : name+"/";
```
```suggestion
            String b = name.isEmpty() ? "" : name + "/";
```

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -248,7 +325,46 @@ public String getFolder() {
+
+    private void handleListFolders(Folder folder, String base, List<String> folders) throws MessagingException  {
+        String name = base + folder.getName();
+        if(!name.isEmpty()) folders.add(name);
```
Why not assign `b` here?

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -248,7 +325,46 @@ public String getFolder() {
+        try {
+            Message[] messages = {message};
+            Folder fromFolder = message.getFolder();
+            Folder toFolder = message.getFolder().getStore().getFolder(toFolderPath);
```
```suggestion
            Folder toFolder = fromFolder.getStore().getFolder(toFolderPath);
```

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/slim/EmailFixture.java:
```diff
@@ -248,7 +325,46 @@ public String getFolder() {
+            fromFolder.copyMessages(messages, toFolder);
+            fromFolder.setFlags(messages, new Flags(Flags.Flag.DELETED), true);
+            fromFolder.expunge();
+            return true;
```
Shouldn't both folders be closed?

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/util/mail/ImapAttachment.java:
```diff
@@ -0,0 +1,23 @@
+    public String getFileName() { return this.fileName; }
+    public InputStream getStream() { return this.stream; }
+
+    public byte[] getBytes() throws IOException  {return IOUtils.toByteArray(this.stream); }
```
Spread across 3 lines

fhoeben reviewed (changes requested):
Looks pretty good functionally. But ...

Commit:
Processed github comments

Commit:
Revisited opening/closing Folders

FIX: Calling setFolders+recieveLastMessage more then X (depended on imap provider) times resulted in execeptions

Commit:
Added:  setFolderReadWrite(int folderReadWrite)

Needed to be able to set write access all the time

fhoeben commented:
I still see some `if` and `for` keywords not followed by a space, some `else {` one a new line (not on the same line as preceding `}`), `if` statements on single line and sometimes without curly braces (instead of spread across 3, with `{`on same line as 'if (' and '}' on new line)

fhoeben commented on src/main/java/nl/hsac/fitnesse/fixture/util/mail/ImapAttachment.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    public byte[] getBytes() throws IOException  {
+        return IOUtils.toByteArray(this.stream);
```
This way each attachment's content can only be accessed once, so one can get it by name of number, but not both. Is that intentional (that doesn't seem like a desirable feature to me)?
Otherwise this class should either store the part itself (so a new inputstream can be retrieved in each cal to `getBytes()`) or the byte array returned by the first call to `getBytes()` should be stored and reused in subsequent calls.

Or am I missing something here?

teunisv replied:
In my test project it saves the same file/attachment fine multiple times

could `return IOUtils.toBufferedInputStream(this.stream);` is realy want though

fhoeben replied:
Good to hear that it already works. That must be because the input stream returned by the part already is some buffered version. Using the method you suggest does sound preferable, as that removes our dependency on implementation detail of the part's input stream.

teunisv replied:
Your correct, fixed it

Commit:
Formating

Commit:
FIX: allow to read getBytes multiple times
